mypy/semanal.py

```diff
@@ -4173,6 +4183,11 @@ def visit_dictionary_comprehension(self, expr: DictionaryComprehension) -> None:
         self.analyze_comp_for_2(expr)
 
     def visit_generator_expr(self, expr: GeneratorExpr) -> None:
+        if True in expr.is_async:
```

Comment:
An async genexp is actually legal in a non-async function: `def f(): return (x async for x in y)`. Could you add a test for that?